### PR TITLE
Add a font supporting more characters

### PR DESCRIPTION
**Pull request by tulior**
Changed the font from Arial to Liberation Sans, which is licensed under the SIL Open Font License.

The tool [fontbuilder](https://github.com/andryblack/fontbuilder/) was used to generate the following character map from the aforementioned font.

2017-04-21-201808_553x532_scrot

This is the whole of printable Extended ASCII.

Also, when matching for the character's id, the type has been changed to unsigned so that charcodes larger than `0x0F` could be used.

### CODE REVIEW

**Comment by doTulio** [comment]
I added a whole lot more characters than you probably would like:
` !"#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\]^_``abcdefghijklmnopqrstuvwxyz{|}~¡¢£¤¥¦§¨©ª«¬®¯°±²³´µ¶·¸¹º»¼½¾¿ÀÁÂÃÄÅÆÇÈÉÊËÌÍÎÏÐÑÒÓÔÕÖ×ØÙÚÛÜÝÞßàáâãäåæçèéêëìíîïðñòóôõö÷øùúûüýþÿЁЄІЇАБВГДЕЖЗИЙКЛМНОПРСТУФХЦЧШЩЪЫЬЭЮЯабвгдежзийклмнопрстуфхцчшщъыьэюяёєіїҐґ`

And the font was changed to Monospace.

Comments?

**Comment by JayFoxRox** [comment]
Please have all information in the first post (= merge your second post into the first post and delete the second post).

Why isn't there a screenshot of the text in action?

---

One font is probably fine for now. It's more important that the exact details used for the fonts creation are documented (e.g. have the fontmap, list of fonts [possibly with download link, so preferably fonts which are easy to license], font size, texture size etc. in the wiki)

Also see https://github.com/andryblack/fontbuilder/issues/19 and 
https://github.com/andryblack/fontbuilder/issues/4 . If those would exist one could just store a config file to generate these fonts more easily. This is beyond the scope of the PR though.

---

I haven't reviewed the code yet. I'll probably do so during the next couple of days. Please remind me if it doesn't happen :)
(Reviews by others appreciated! @anodium ?)

File: main.cpp [hunk 1]
```diff
@@ -66,7 +66,8 @@ int main(int argc, char* argv[]) {
 
   std::shared_ptr<Texture> texture = std::make_shared<Texture>("data/font/main");
 
-  std::shared_ptr<Text> text = std::make_shared<Text>("data/font/arial_black_regular_65");
+  //std::shared_ptr<Text> text = std::make_shared<Text>("data/font/arial_black_regular_65");
```
Comment (anodium) [inline main.cpp]: Why was this commented out instead of edited directly?

File: text.cpp [hunk 2]
```diff
@@ -91,7 +91,7 @@ class Text {
       }
 
       auto isMatchingSymbol = [&](const Symbol& symbol) -> bool {
-        return symbol.id == static_cast<uint32_t>(c);
+          return symbol.id == static_cast<unsigned char>(c);
```
Comment (anodium) [inline text.cpp]: What purpose does changing the type have?

File: main.cpp [hunk 3]
```diff
@@ -466,7 +467,17 @@ int main(int argc, char* argv[]) {
 
     // Test text
     glColor3f(0.1f, 0.2f, 0.5f);
-    text->draw(std::string("boats: ") + std::to_string(boats.size()) + std::string("\nsparks: ") + std::to_string(sparks.size()) + std::string("\nus/F: ") + std::to_string(uspf.count()));
+
```
Comment (anodium) [inline main.cpp]: I see ninja refactoring. :eyes:

If you're going to ninja @JayFoxRox, at least clean up your whitespace. :wink:

**Review by anodium — COMMENTED** [review]
Why is Arial Black still in the codebase, if Liberation Mono does its job and more?

File: text.cpp (hunk 2 above)
Comment (JayFoxRox) [inline text.cpp]: Whitespace change not necessary

Comment (JayFoxRox) [reply text.cpp]: Yes. Also if the type is unsigned here it should probably also be like that in the zengl struct (which is currently `int32_t`).
However, I don't know how unicode works in C++ and if zengl files even use unicode (which I kind of hope).

**Comment by tulior** [comment]
Pushed new commit. Also, the list of supported characters has been greatly reduced, since we are not doing Unicode. The new character set is shown below.
2017-04-21-201808_553x532_scrot

File: main.cpp [hunk 4]
```diff
@@ -467,17 +466,8 @@ int main(int argc, char* argv[]) {
...
-    
-    std::string debuggingText =
-        std::string("boats: ") +
-            std::to_string(boats.size()) +
-        std::string("\nsparks: ") +
-            std::to_string(sparks.size()) +
-        std::string("\n\xB5/F: ") + //'\xB5' being the micro symbol
-            std::to_string(uspf.count());
-    text->draw(debuggingText);
+    //'\xB5' = µ = micro sign (in case your editor's font can't handle it)
```
Comment (JayFoxRox) [inline main.cpp]: nit: Ideally this should have been in a variable: char symbolMicro = '\xB5'; or something.

File: text.cpp [hunk 5]
```diff
@@ -91,7 +91,16 @@ class Text {
       }
 
       auto isMatchingSymbol = [&](const Symbol& symbol) -> bool {
-          return symbol.id == static_cast<unsigned char>(c);
+        /* c is probably char_t here, so ...
+          casting first to unsigned char is necessary, because if we
+          cast to int32_t directly, the machine bit extends, meaning
```
Comment (JayFoxRox) [inline text.cpp]: *sign extends

File: text.cpp [hunk 6]
```diff
@@ -91,7 +91,16 @@ class Text {
       }
 
       auto isMatchingSymbol = [&](const Symbol& symbol) -> bool {
-          return symbol.id == static_cast<unsigned char>(c);
+        /* c is probably char_t here, so ...
```
Comment (JayFoxRox) [inline text.cpp]: -probably

File: text.cpp [hunk 7]
```diff
@@ -91,7 +91,16 @@ class Text {
...
+          casting first to unsigned char is necessary, because if we
+          cast to int32_t directly, the machine bit extends, meaning
+          that the match would fail because of the type's sign.
+          we could also do uint32_t, since I think zengl supports utf-32
+          but I don't think that's necessary because we are not support 
+          unicode anyway (i.e. we never go beyond 0xFF)
```
Comment (JayFoxRox) [inline text.cpp]: depending on what zengl uses we should fix the zengl struct and also this cast. I stole my zengl struct from the font tool source and changed some datatypes to stdint iirc.

**Comment by JayFoxRox** [comment]
Only some smaller issues left, I will test this later and hopefully we can integrate it in a couple of hours

**Comment by tulior** [comment]
Hopefully everything is OK now.
This is what the last version looks like.
2017-04-23-123525_213x143_scrot

File: text.cpp [hunk 8]
```diff
@@ -91,7 +91,16 @@ class Text {
       }
 
       auto isMatchingSymbol = [&](const Symbol& symbol) -> bool {
-        return symbol.id == static_cast<uint32_t>(c);
+        /* c is char_t here, so ...
```
Comment (JayFoxRox) [inline text.cpp]: isn't it just a `char` ?

File: text.cpp [hunk 9]
```diff
@@ -91,7 +91,16 @@ class Text {
...
+          casting first to unsigned char is necessary, because if we
+          cast to int32_t directly, the machine sign extends, meaning
+          that the match would fail because of the type's sign.
```
Comment (JayFoxRox) [inline text.cpp]: -type's

File: text.cpp [hunk 10]
```diff
@@ -91,7 +91,16 @@ class Text {
...
+          we could also do uint32_t, since I think zengl supports utf-32
+          but I don't think that's necessary because we are not support 
+          unicode anyway (i.e. we never go beyond 0xFF)
+        */
+        unsigned char chUnsigned = static_cast<unsigned char>(c);
+        return symbol.id == static_cast<int32_t>(chUnsigned);
```
Comment (JayFoxRox) [inline text.cpp]: symbol_id is uint32_t, so this should be uint32_t too

File: text.cpp [hunk 11]
```diff
@@ -91,7 +91,16 @@ class Text {
       }
 
       auto isMatchingSymbol = [&](const Symbol& symbol) -> bool {
-        return symbol.id == static_cast<uint32_t>(c);
+        /* c is char_t here, so ...
+          casting first to unsigned char is necessary, because if we
+          cast to int32_t directly, the machine sign extends, meaning
+          that the match would fail because of the type's sign.
+          we could also do uint32_t, since I think zengl supports utf-32
```
Comment (JayFoxRox) [inline text.cpp]: We already do a 32 bit conversion here, so I'm not sure why you say "we could" - we already do. (also we can't - c is a char and we won't change that. So information has already been lost)

**Comment by JayFoxRox** [comment]
Okay, please fix what's mentioned in the remaining comments, squash up and make sure to add the character map to the wiki.

Reviewed and tested: LGTM aside from what's mentioned

File: text.cpp (hunk 8 above)
Comment (JayFoxRox) [reply text.cpp]: forgot about this one? @doTulio 
It's a nit, so if you don't change it I'll just merge anyway in about 10 minutes or so.

**Comment by JayFoxRox** [comment]
Thanks!

(Please add the charmap and other settings you used + a link to the font (as ttf or w/e) to the wiki)

**Pull request merged**